### PR TITLE
Add coln-batch, first draft of the batch query engine

### PR DESCRIPTION
**Pull request by szlangini**
First draft of the batch query engine, as a standalone crate. 
  Wired it into `just check-rust`.

  **Goal**: a mockup scenario for the read path, end to end. We generate
  join workloads (a cyclic triangle and an acyclic e-matching pattern),
  store them as Arrow files, load and sort them into `SortedTable`s, and
  hand those to the engine, where two join algorithms answer conjunctive
  queries over the data and return a result relation.

  **What's in it:**
  
  - conjunctive query representation (atoms + head, mirrors FLIR's rule
    shape so the later adapter stays mechanical) - will supposedly migrate later to FLIR.
  - two executors, a binary hash-join chain and a worst-case-optimal
    generic join
  - the `SortedTable` trait, the only way the engine reads data. Backed
    by an in-memory Arrow implementation for now, real storage plugs in
    behind it later. `check_contract` validates any implementation
    against the contract.
  - deterministic workload generators and Arrow IPC round-trip (save and load) for test
    data, plus a brute-force oracle (join by hand)

  **Testing**: the executors are differential-tested against each other and
  against the oracle, on hand-written edge cases, fixture workloads, and
  250 randomized query shapes per run. `--include-ignored` runs 10M-row
  versions in release mode and prints per-executor timings. `just check`
  in the package mirrors CI.

  To see it **run**: `cargo run -p coln-batch --example smoke` (three edges,
  output checkable by hand), or `--example demo --release` for the full
  pipeline, which ends in "All checks passed."

  Out of scope here: performance (marked TODO(perf) in the code),
  recursion (ready on a follow-up branch), FLIR intake, optimizer.

### CODE REVIEW

**Comment by szlangini** [comment]
Seems like the failing gate is not due to my commits:

`examples/sync-demo/src/generated/Graph.ts: no license identifier
examples/sync-demo/src/generated/Graph.ts: no copyright notice
examples/sync-demo/src/generated/GraphRealm.ts: no license identifier
examples/sync-demo/src/generated/GraphRealm.ts: no copyright notice
error: recipe `check-licenses` failed on line 24 with exit code 123
Error: Process completed with exit code 123.`

File: packages/coln-batch/src/table.rs [hunk 1]
```diff
@@ -0,0 +1,325 @@
...
+    /// Number of rows.
+    fn len(&self) -> usize;
+
+    fn is_empty(&self) -> bool {
+        self.len() == 0
+    }
+
+    /// The column order rows are sorted by: a permutation of `0..arity()`.
+    /// `sort_order()[0]` is the major sort column.
+    fn sort_order(&self) -> &[ColId];
```
Comment (lstwn) [inline packages/coln-batch/src/table.rs:42]: A table can have one or more (possibly compound) primary keys. I bet this is useful to know. Maybe we should add a fn for that, too? Something like:

```
fn primary_keys(&self) -> &[&[ColId]]
```

File: packages/coln-batch/src/table.rs [hunk 2]
```diff
@@ -0,0 +1,325 @@
...
+
+    /// Cell access. `row` is a position in *sorted* order (`0..len()`);
+    /// `col` is a column id in *schema* order.
+    fn value(&self, row: usize, col: ColId) -> u64;
```
Comment (lstwn) [inline packages/coln-batch/src/table.rs]: So for now we only support tables with `u64` entries?

File: packages/coln-batch/src/table.rs [hunk 3]
```diff
@@ -0,0 +1,325 @@
...
+use std::cmp::Ordering;
+use std::ops::Range;
+
+use anyhow::{Result, bail};
+use arrow::record_batch::RecordBatch;
+
+use crate::relation::Relation;
+
+/// Column id, in *schema* order (position in the relation's column list).
+pub type ColId = usize;
```
Comment (lstwn) [inline packages/coln-batch/src/table.rs:23]: Why not add a `RowIdx`/`RowId` type alias, too?

File: packages/coln-batch/src/table.rs (hunk 2 above)
Comment (szlangini) [reply packages/coln-batch/src/table.rs]: yes, for the demo only. will add more types to support later

File: packages/coln-batch/src/table.rs (hunk 3 above)
Comment (szlangini) [reply packages/coln-batch/src/table.rs:23]: agree
